### DISCUSSION TITLE
Getting "nitro/cache-driver.js" is imported by "virtual:#internal/nitro/virtual/storage", but could not be resolved – treating it as an external dependency.

### DISCUSSION THREAD

**Discussion by nachoadjust** [nuxt/nuxt#26622] [Questions]
Since the release of 3.11.1 I started getting this warn, and I don't know how to fix it:
`[warn] "file:///vercel/path0/node_modules/nuxt/dist/core/runtime/nitro/cache-driver.js" is imported by "virtual:#internal/nitro/virtual/storage", but could not be resolved – treating it as an external dependency.`

In this case I was building for vercel, but I also get this warn when building locally with preset: node-server.

If anyone could point me in the correct direction to fix it I'd be very thankful!
Thanks!!

**Comment by victorlmneves** [comment]
@nachoadjust 
Same happening to me when running the build locally
Are setting routeRules?
I start seeing while working on a solution to generate plain static HTML pages just for a couple of routes

**Reply by RollingTL** [reply]
Same happening when i run 
`npx nuxi generate`
And i have
`  ssr: true`
in nuxt.config.ts
[3:55:55 PM]  WARN  "file:///...../node_modules/nuxt/dist/core/runtime/nitro/cache-driver.js" is imported by "virtual:#internal/nitro/virtual/storage", but could not be resolved – treating it as an external dependency.

**Comment by RollingTL** [comment]
Updated
` nuxt                ^3.11.1  →  ^3.11.2`
No warning when npx nuxi generate.
So it look like the problem solved.

**Comment by vanminhquangtri** [comment]
Hello all, I have same issue, tried to update version nuxt to 3.11.2 but warning still exists. Thanks

**Reply by Leo-Felde** [reply]
Same for me, I'm going crazy trying to figure out what might be causing this (╯'□')╯︵ ┻━┻

**Reply by Leo-Felde** [reply]
Yeah just dont use ssr: true and it should work.....
┬─┬﻿ ノ( ゜-゜ノ)

**Comment by guillaumeprevost** [comment]
I get this same warning when running `nuxt build` 
Versions : Nuxt 3.11.2, Nitro 2.9.6, vite v5.2.8

```
[14:47:42]  WARN  "file:///C:/dev/................/node_modules/nuxt/dist/core/runtime/nitro/cache-driver.js" is imported by "virtual:#internal/nitro/virtual/storage", but could not be resolved – treating it as an external dependency.
```

Additionally, and am not sure the 2 issues are linked, the .output folder remains with 2 EMPTY folders (public & server) : nothing is copied from .nuxt, and that prevents running the web app (see full logs & details in SO issue I created about that other issue : https://stackoverflow.com/questions/78366245/nuxt3-build-output-folder-remains-empty-nothing-copied-from-nuxt)

**Comment by Atulin** [comment]
I'm getting the same error on Nuxt 3.11.2, the build script fails, too

```
ℹ ✓ built in 1.30s                                                                                                                                                          7:36:10   
✔ Server built in 1321ms                                                                                                                                                    7:36:10   
ℹ Initializing prerenderer                                                                                                                                            nitro 7:36:10

[7:36:12]  WARN  "file:///G:/[...]/node_modules/nuxt/dist/core/runtime/nitro/cache-driver.js" is imported by "virtual:#internal/nitro/virtual/storage", but 
could not be resolved – treating it as an external dependency.

ℹ Prerendering 3 initial routes with crawler                                                                                                                          nitro 7:36:12
  ├─ /200.html (10139ms)                                                                                                                                               nitro 7:36:22
  ├─ /404.html (10139ms)                                                                                                                                               nitro 7:36:22
  ├─ / (10209ms)                                                                                                                                                       nitro 7:36:23
  ├─ /_payload.json?fa67fc4d-a859-4f6e-975f-51b63b014448 (20ms) (skipped)                                                                                              nitro 7:36:33
error: script "generate" exited with code 5
```

Updating to `3.12.2` fixed the issue, though.

**Comment by balajiksk** [comment]
[5:39:12 pm]  WARN  "file:///C:/nuxt/app/csk/node_modules/nuxt/dist/core/runtime/nitro/cache-driver.js" is imported by "virtual:#internal/nitro/virtual/storage", but could not be resolved – treating it as an external dependency.

I got same warning. because of this error `useCookie` destroyed while page reload in github server
in localhost its working fine.
I using 
Nuxt 3.12.2 with Nitro 2.9.6
vite v5.3.1

**Comment by morovinger** [comment]
same here

**Comment by shlomi93** [comment]
Same here..
Solution?

**Comment by coderino42** [comment]
I get this issue when I build while the dev server is running. I've seen this on several versions and am still seeing it on 3.12.4.  I have been able to run them both successfully before on other projects running older Nuxt versions. The dev server does restart in that case, but the build completes. Hope this helps.

**Comment by 0xMaddyy** [comment]
Boa tarde senhores, eu acesse esse link para procurar como resolver o meu problema, desculpe-me, mas como eu sou do Brasil, vou escrever em Português, mesmo o Google traduz, risos.
No meu arquivo de configuração, nuxt.config.ts, eu adicionei o seguinte código:
```
export default defineNuxtConfig({
  ssr: true,
  nitro: {
    preset: 'node-server'
  },
...
```
No meu projeto que tem Nuxt3 com Quasar, funcionou tranquilo, e creio que funcionará pra vocês.
Um grande abraço e sucesso a todos!

**Reply by sugoidesune** [reply]
this fixed it for me! well done.

**Reply by Mgoeloe** [reply]
> Boa tarde senhores, eu acesse esse link para procurar como resolver o meu problema, desculpe-me, mas como eu sou do Brasil, vou escrever em Português, mesmo o Google traduz, risos. No meu arquivo de configuração, nuxt.config.ts, eu adicionei o seguinte código:
> 
> ```
> export default defineNuxtConfig({
>   ssr: true,
>   nitro: {
>     preset: 'node-server'
>   },
> ...
> ```
> 
> No meu projeto que tem Nuxt3 com Quasar, funcionou tranquilo, e creio que funcionará pra vocês. Um grande abraço e sucesso a todos!



Didn't solve the issue for me

**Reply by 0xMaddyy** [reply]
You have other problems too!
View this project and test.
https://github.com/ricavalier/quasar-nuxt

**Comment by petacreepers23** [comment]
Same issue here on latest as of today

**Comment by itsDonya** [comment]
Same here

**Comment by NOVACOAX** [comment]
### Clean node_modules and Reinstall Dependencies
Sometimes dependency resolution issues can arise from cached or old packages. Try removing node_modules and reinstalling your dependencies:

```
rm -rf node_modules package-lock.json yarn.lock
npm install
```

or if using yarn:

```
rm -rf node_modules yarn.lock
yarn install
```

**Comment by NOVACOAX** [comment]
I don't know if this will help anyone, but I had almost the same issue while trying to host on `firebase`: I looked up the [Nitro website](https://nitro.unjs.io/deploy). 

They say to install some dependencies:
```
npm i firebase-admin firebase-functions firebase-functions-test
```
This information was not on the Nuxt Documentation page.

**Comment by simeon9696** [comment]
November 24th still having this issue. Don't know where to start but this is my package.json

```
{
  "name": "nuxt-app",
  "private": true,
  "type": "module",
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev --host",
    "generate": "nuxt generate",
    "preview": "nuxt preview",
    "postinstall": "nuxt prepare"
  },
  "dependencies": {
    "@nauverse/make-url": "^1.10.17",
    "@pinia/nuxt": "^0.7.0",
    "@quasar/extras": "^1.16.11",
    "firebase": "^11.0.2",
    "nuxt": "^3.11.2",
    "nuxt-quasar-ui": "^2.1.2",
    "pinia": "^2.1.7",
    "pinia-plugin-persistedstate": "^4.1.3",
    "quasar": "^2.16.4",
    "slug": "^10.0.0",
    "v-viewer": "^3.0.20",
    "viewerjs": "^1.11.6",
    "vue": "^3.4.21",
    "vue-router": "^4.3.0"
  },
  "devDependencies": {
    "@types/slug": "^5.0.9"
  },
  "optionalDependencies": {
    "@rollup/rollup-linux-x64-gnu": "*"
  }
}

```

**Comment by Yakima-Teng** [comment]
I have this issue too. I am crazy about upgrading nuxt. Through I can make it work, but it take some time. I am not upgrading from nuxt2 to nuxt3, but from nuxt3 to nuxt3.

Here is my package.json:

```json
{
  "name": "ols-viewer",
  "private": true,
  "type": "module",
  "scripts": {
    "ready": "npm install",
    "lint": "eslint --fix .",
    "dev:test": "cross-env NODE_ENV=development LOG_SQL=1 nuxt dev --dotenv .env.test",
    "dev:production": "cross-env NODE_ENV=development LOG_SQL=1 nuxt dev --dotenv .env.production",
    "build:test": "nuxt build --dotenv .env.test && node build/update-output-package.cjs",
    "build:production": "nuxt build --dotenv .env.production && node build/update-output-package.cjs",
    "build": "npm run build:production",
    "start": "cd ./.output/server && npm install && cross-env NODE_ENV=production PORT=3000 HOST=0.0.0.0 pm2-runtime start ./.output/server/index.mjs --name ols-viewer",
    "postinstall": "nuxt prepare"
  },
  "devDependencies": {
    "@element-plus/nuxt": "1.1.1",
    "@nuxt/devtools": "latest",
    "@nuxt/eslint": "0.7.5",
    "@nuxtjs/i18n": "9.2.1",
    "@nuxtjs/supabase": "1.4.5",
    "@types/crypto-js": "4.2.1",
    "@types/jsonwebtoken": "9.0.5",
    "@types/nodemailer": "6.4.14",
    "@types/uuid": "9.0.7",
    "cross-env": "7.0.3",
    "dotenv": "16.4.1",
    "element-plus": "2.9.3",
    "home": "2.0.0",
    "node-ssh": "13.1.0",
    "nuxt": "3.15.2",
    "nuxt-gtag": "3.0.2",
    "pm2": "5.4.3",
    "sass": "1.85.1",
    "typescript": "5.3.3",
    "vue": "^3.3.8",
    "vue-router": "^4.2.5"
  },
  "dependencies": {
    "@nuxt/content": "2.11.0",
    "@nuxt/kit": "3.15.2",
    "@pinia/nuxt": "0.5.1",
    "@supabase/supabase-js": "2.48.0",
    "axios": "1.6.2",
    "cheerio": "1.0.0-rc.12",
    "crypto-js": "4.2.0",
    "pinia": "2.1.7",
    "utils-daily": "1.0.1",
    "uuid": "9.0.1"
  }
}
```

**Comment by simonmaass** [comment]
i am still having this in nuxt 4 - anyone found a solution?

**Reply by Ercilan** [reply]
https://github.com/nuxt/nuxt/issues/27424#issuecomment-2144689117

**Comment by cmdragon** [comment]
I also encountered this issue while building an application with nuxt 3.20.1.
